File: Python/Binary Tree/reverse_level_order_traversal.py
# Problem link - https://www.geeksforgeeks.org/problems/reverse-level-order-traversal/1


from abc import ABC, abstractmethod


class TreeNode:
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None


class DataNode:
    def __init__(self, data, level):
        self.data = data
        self.level = level


class Node:
    def __init__(self, node: DataNode):
        self.data = node
        self.next = None


class CollectionInterface(ABC):
    def __init__(self):
        self.head = self.tail = None
        self.length = 0

    @abstractmethod
    def push(self, node: DataNode):
        pass

    @abstractmethod
    def pop(self):
        pass

    def is_empty(self):
        return self.length == 0


class Stack(CollectionInterface):
    def pop(self):
        if self.is_empty():
            return
        item = self.head.data
        node = self.head
        self.head = self.head.next
        del node
        self.length -= 1
        return item

    def push(self, x: DataNode):
        node = Node(x)
        if self.is_empty():
            self.head = self.tail = node
        else:
            node.next = self.head
            self.head = node
        self.length += 1


class Queue:
    def __init__(self):
        self.s1 = Stack()
        self.s2 = Stack()

    def push(self, x: DataNode):
        while not self.s1.is_empty():
            self.s2.push(self.s1.pop())
        self.s1.push(x)
        while not self.s2.is_empty():
            self.s1.push(self.s2.pop())

    def pop(self):
        return self.s1.pop()

    def is_empty(self):
        return self.s1.is_empty()


class Solution:
    @staticmethod
    def reverse_level_order(root: TreeNode):
        # create blank stack and queue data structures
        stack = Stack()
        queue = Queue()

        # push the root node with level 0.
        queue.push(DataNode(root, 0))

        # do the BFS traversal till the queue is not empty.
        while not queue.is_empty():
            # pop the current data node from queue.
            node = queue.pop()
            # push the data node to the stack.
            stack.push(node)

            # now first check the right child instead of left and push it to the queue with +1 level value.
            # We check right first because in the reversed order we want to see left child first when popping out from
            # the stack.
            if node.data.right is not None:
                queue.push(DataNode(node.data.right, node.level + 1))
            if node.data.left is not None:
                queue.push(DataNode(node.data.left, node.level + 1))

        # once stack is populated, pop from it until it becomes empty and store the integer data in the result list.
        result = []
        while not stack.is_empty():
            result.append(stack.pop().data.data)

        # the result list finally contains the reversed level order traversal.
        return result


# Example 1
one, two, three = TreeNode(1), TreeNode(2), TreeNode(3)
one.left = two
one.right = three
print(Solution.reverse_level_order(one))

# Example 2
ten, twenty, thirty, fourty, sixty = TreeNode(10), TreeNode(20), TreeNode(30), TreeNode(40), TreeNode(60)
ten.left = twenty
twenty.left = fourty
ten.right = thirty
twenty.right = sixty
print(Solution.reverse_level_order(ten))

File: Java/BinaryTrees/Problem2/Reverse Level Order Traversal.java
// Problem link - https://www.geeksforgeeks.org/problems/reverse-level-order-traversal/1

package BinaryTrees.Problem2;


import java.util.ArrayList;
import java.util.List;

class QueueNode<T> {
    private T data;
    private QueueNode<T> next;

    public QueueNode(T data) {
        this.data = data;
        this.next = null;
    }

    public QueueNode<T> getNext() {
        return next;
    }

    public void setNext(QueueNode<T> next) {
        this.next = next;
    }

    public T getData() {
        return data;
    }

    public void setData(T data) {
        this.data = data;
    }
}


class Queue<T> {
    private QueueNode<T> head;
    private QueueNode<T> tail;
    private Integer length;

    public Queue() {
        this.head = this.tail = null;
        this.length = 0;
    }

    public boolean isEmpty() {
        return this.length.equals(0);
    }

    public void push(T x) {
        QueueNode<T> node = new QueueNode<>(x);
        if (isEmpty()) {
            this.head = this.tail = node;
        } else {
            this.tail.setNext(node);
            this.tail = node;
        }
        this.length += 1;
    }

    public T pop() {
        if (isEmpty()) {
            return null;
        }
        T item = this.head.getData();
        this.head = this.head.getNext();
        this.length -= 1;
        return item;
    }
}


class Node<T> {
    private T data;
    private Node<T> left;
    private Node<T> right;

    public Node(T data) {
        this.data = data;
        this.left = this.right = null;
    }

    public Node<T> getLeft() {
        return left;
    }

    public void setLeft(Node<T> left) {
        this.left = left;
    }

    public Node<T> getRight() {
        return right;
    }

    public void setRight(Node<T> right) {
        this.right = right;
    }

    public T getData() {
        return data;
    }
}


class Solution {
    public static void main(String[] args) {
        // Example 1
        Node<Integer> n1 = new Node<>(1);
        Node<Integer> n2 = new Node<>(2);
        Node<Integer> n3 = new Node<>(3);
        n1.setLeft(n3);
        n1.setRight(n2);
        System.out.println(getReverseOrderTraversal(n1));

        // Example 2
        Node<Integer> n10 = new Node<>(10);
        Node<Integer> n20 = new Node<>(20);
        Node<Integer> n30 = new Node<>(30);
        Node<Integer> n40 = new Node<>(40);
        Node<Integer> n60 = new Node<>(60);
        n10.setLeft(n20);
        n10.setRight(n30);
        n20.setLeft(n40);
        n20.setRight(n60);
        System.out.println(getReverseOrderTraversal(n10));
    }

    public static <T> List<T> getReverseOrderTraversal(Node<T> root) {
        /**
         * Time complexity is O(n) and space complexity is O(n).
         */

        // define a blank queue and push root node into it.
        Queue<Node<T>> queue = new Queue<>();
        queue.push(root);

        // define a result list.
        List<T> result = new ArrayList<>();

        // typical BFS...
        while (!queue.isEmpty()) {
            Node<T> node = queue.pop();

            // add the popped node's data into the result list.
            result.add(node.getData());

            // add right node first, then left node into the queue.
            // We check right first because in the reversed order we want to see left child first.
            if (node.getRight() != null) {
                queue.push(node.getRight());
            }
            if (node.getLeft() != null) {
                queue.push(node.getLeft());
            }
        }

        // return the reversed result for the final answer in O(n) time.
        return result.reversed();
    }
}